zebrad/src/components/miner.rs

```diff
@@ -0,0 +1,527 @@
+        );
+        abort_handles.push(solver.abort_handle());
+
+        mining_solvers.push(solver.wait_for_panics());
+    }
+
+    // These tasks run forever unless there is a fatal error or shutdown.
+    // When that happens, the first task to error returns, and the other JoinHandle futures are
+    // cancelled.
+    let first_result;
+    select! {
+        result = template_generator => { first_result = result; }
+        result = mining_solvers.try_next() => {
+            first_result = result
+                .transpose()
+                .expect("stream never teminates because there is at least one solver task");
+        }
+    }
+
+    // But the spawned async tasks keep running, so we need to abort them here.
+    for abort_handle in abort_handles {
+        abort_handle.abort();
+    }
+
+    // Any spawned blocking threads will keep running. When this task returns and drops the
+    // `template_sender`, it cancels all the spawned miner threads. This works because we've
+    // aborted the `template_generator` task, which owns the `template_sender`. (And it doesn't
+    // spawn any blocking threads.)
+    first_result
+}
+
+/// Generates block templates using `rpc`, and sends them to mining threads using `template_sender`.
+#[instrument(skip(rpc, template_sender))]
+pub async fn generate_block_templates<
+    Mempool,
+    State,
+    Tip,
+    BlockVerifierRouter,
+    SyncStatus,
+    AddressBook,
+>(
+    rpc: GetBlockTemplateRpcImpl<Mempool, State, Tip, BlockVerifierRouter, SyncStatus, AddressBook>,
+    template_sender: watch::Sender<Option<Arc<Block>>>,
+) -> Result<(), Report>
+where
+    Mempool: Service<
+            mempool::Request,
+            Response = mempool::Response,
+            Error = zebra_node_services::BoxError,
+        > + Clone
+        + Send
+        + Sync
+        + 'static,
+    Mempool::Future: Send,
+    State: Service<
+            zebra_state::ReadRequest,
+            Response = zebra_state::ReadResponse,
+            Error = zebra_state::BoxError,
+        > + Clone
+        + Send
+        + Sync
+        + 'static,
+    <State as Service<zebra_state::ReadRequest>>::Future: Send,
+    Tip: ChainTip + Clone + Send + Sync + 'static,
+    BlockVerifierRouter: Service<zebra_consensus::Request, Response = block::Hash, Error = zebra_consensus::BoxError>
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+    <BlockVerifierRouter as Service<zebra_consensus::Request>>::Future: Send,
+    SyncStatus: ChainSyncStatus + Clone + Send + Sync + 'static,
+    AddressBook: AddressBookPeers + Clone + Send + Sync + 'static,
+{
+    // Pass the correct arguments, even if Zebra currently ignores them.
+    let mut parameters = get_block_template::JsonParameters {
+        mode: Template,
+        data: None,
+        capabilities: vec![LongPoll, CoinbaseTxn],
+        long_poll_id: None,
+        _work_id: None,
```

Comment:
Subjective/Nitpick:

```suggestion
        capabilities: vec![LongPoll, CoinbaseTxn],
        ..Default::default()
```

Reply:
I think it's clearer if we're explicit about the `mode` (because we want a template), and `long_poll_id` (because we change it later).

I would be ok with replacing the other fields with defaults if you submitted another PR.